src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,380 @@
 ---
 title: Fixing type promotion failures
 description: Solutions for cases where you know more about a field's type than Dart can determine.
-toc: false
 ---
 
-{{site.alert.warn}}
-  **This page is under construction**
-  ([issue #2940][]).
-{{site.alert.end}}
-
-This page will have information to help you understand
+This page has information to help you understand
 why type promotion failures occur,
-and give tips on how to fix them.
-For example, consider the following code:
+with tips on how to fix them.
+For background information,
+see [Working with nullable fields][ns-fields],
+a section in [Understanding null safety][].
+
+[issue #2940]: https://github.com/dart-lang/site-www/issues/2940
+[ns-fields]: /null-safety/understanding-null-safety#working-with-nullable-fields
+[Understanding null safety]: /null-safety/understanding-null-safety
+
+
+## Only local variables can be promoted {#property-or-this}
 
-```dart
+**The cause:**
+You're trying to promote a property or `this`,
+but only local variables can be promoted.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
 class C {
   int? i;                  // (1)
-  f() {
+  void f() {
     if (i == null) return;
-    i.isEven;              // (2)
+    i.isEven;              // (2) ERROR
```

Comment:
I'm guessing you got this example from me, because I write a lot test cases using `int` and `isEven` like this.  I wonder if it would be better to try to come up with less contrived examples since this is user-facing documentation.  I'm spitballing here, but maybe something like this?

```dart
class Order {
  DateTime? shipDate; // (1)
  void checkForShipmentDelay() {
    if (shipDate == null) return;
    if (DateTime.now().difference(shipDate) > Duration(days: 7)) { // (2) ERROR
      print('Shipment is late');
    }
  }
}
```

Unfortunately, I don't really have good ideas for making the other examples less contrived (they are less common than field promotion, so it's harder to invent concrete examples of where they might occur).  Maybe it's better to stick with `int` and `isEven` for now, and wait and see if anyone complains?  I don't know.

At the very least, consider changing this line to `print(i.isEven);` so that it doesn't just look like a statement with no effect 😃

Reply:
Using print() would be nice. :) I do like keeping it simple, and contrived is fine as long as it helps readability/understandability.

Reply:
I just noticed a weird URL when I put the code in DartPad:

![image](https://user-images.githubusercontent.com/2164483/120569223-544bdc80-c3ca-11eb-8b48-18fe71ef31ce.png)

Reply:
(It should be http://dart.dev/go/non-promo-property.)

Reply:
This renders properly for me in WebStorm, so perhaps it's an issue with how DartPad sends or displays the diagnostic message.

Seems like it's received liked that from dart-services:
![image](https://user-images.githubusercontent.com/18372958/120698771-1e176700-c475-11eb-9707-0e4d8c7a8f9b.png)

cc @domesticmouse